HeterogeneousCore/AlpakaInterface/test/alpaka/testAtomicPairCounter.dev.cc

```diff
@@ -0,0 +1,110 @@
+#include <cassert>
+#include <iostream>
+
+#define CATCH_CONFIG_MAIN
+#include <catch.hpp>
+
+#include "HeterogeneousCore/AlpakaInterface/interface/config.h"
+#include "HeterogeneousCore/AlpakaInterface/interface/memory.h"
+#include "HeterogeneousCore/AlpakaInterface/interface/workdivision.h"
+#include "HeterogeneousCore/AlpakaInterface/interface/AtomicPairCounter.h"
+
+using namespace cms::alpakatools;
+using namespace ALPAKA_ACCELERATOR_NAMESPACE;
+
+static constexpr auto s_tag = "[" ALPAKA_TYPE_ALIAS_NAME(alpakaTestAtomicPair) "]";
+
+struct update {
+  template <typename TAcc>
+  ALPAKA_FN_ACC void operator()(
+      const TAcc &acc, AtomicPairCounter *dc, uint32_t *ind, uint32_t *cont, uint32_t n) const {
+    for_each_element_in_grid_strided(acc, n, [&](uint32_t i) {
+      auto m = i % 11;
+      m = m % 6 + 1;  // max 6, no 0
+      auto c = dc->inc_add(acc, m);
+      assert(c.first < n);
+      ind[c.first] = c.second;
+      for (uint32_t j = c.second; j < c.second + m; ++j)
+        cont[j] = i;
+    });
+  }
+};
+
+struct finalize {
+  template <typename TAcc>
+  ALPAKA_FN_ACC void operator()(
+      const TAcc &acc, AtomicPairCounter const *dc, uint32_t *ind, uint32_t *cont, uint32_t n) const {
+    assert(dc->get().first == n);
+    ind[n] = dc->get().second;
+  }
+};
+
+TEST_CASE("Standard checks of " ALPAKA_TYPE_ALIAS_NAME(alpakaTestAtomicPair), s_tag) {
+  SECTION("AtomicPairCounter") {
+    auto const &devices = cms::alpakatools::devices<Platform>();
+    if (devices.empty()) {
+      std::cout << "No devices available on the platform " << EDM_STRINGIZE(ALPAKA_ACCELERATOR_NAMESPACE)
+                << ", the test will be skipped.\n";
+      return;
+    }
```

Comment:
Given the somewhat recent changes to how scram runs the unit tests for the alpaka backends (i.e. e.g. cuda tests are ran only if `cudaIsEnabled` program returned `0`), I'd argue this way of skipping the test in absence of device is not needed, and it would actually be better if the test fails in absence of the device(s).

Reply:
Without this section, we still get a success (but with no message) as the tests run inside this loop:

```C++
for (auto const &device : devices) {
```
Another option would be to explicitly fail (```assert(!devices.empty())``` ?). Is this a more standard behavior?

Reply:
I think it would be good for these tests to require a device (but I'd use catch2's `REQUIRE()` rather than `assert()`).

Reply:
> Given the somewhat recent changes to how scram runs the unit tests for the alpaka backends (i.e. e.g. cuda tests are ran only if `cudaIsEnabled` program returned `0`), I'd argue this way of skipping the test in absence of device is not needed, and it would actually be better if the test fails in absence of the device(s).

How do these changes work?
Should we update all existing tests accordingly?

Reply:
> > Given the somewhat recent changes to how scram runs the unit tests for the alpaka backends (i.e. e.g. cuda tests are ran only if `cudaIsEnabled` program returned `0`), I'd argue this way of skipping the test in absence of device is not needed, and it would actually be better if the test fails in absence of the device(s).
> 
> How do these changes work?

The behavior is described in https://github.com/cms-sw/cmssw/issues/41340#issuecomment-1514867660

> Should we update all existing tests accordingly?

I already did privately for `HeterogeneousCore/AlpakaTest` last week (just need to rebase now that the preceding PRs got merged).

About direct CUDA I'm not sure. Ideally yes, but is it worth of the effort?